---
title: |
  | Differential abundance -- <br> Salmonella project metabarcoding data
pagetitle: "Differential abundance"
author: "Jelmer Poelstra (poelstra.1@osu.edu), MCIC Wooster"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: cerulean
    highlight: kate
    toc: true
    toc_float: true
    fig_caption: true
    anchor_sections: true
    df_print: kable
    css: html_page.css
bibliography: salmonella.bib
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  eval = TRUE,
  cache = FALSE,
  warning = FALSE,
  message = FALSE,
  out.width = "80%"
)
knitr::opts_knit$set(root.dir = here::here(),
                     output.dir = here::here("reports"))
```

```{r}
## Load packages
if (!"pacman" %in% installed.packages()) install.packages("pacman")
packages <- c("tidyverse", "here",
              "phyloseq", "microbiome",
              "DESeq2", "ALDEx2",
              "pheatmap", "patchwork", "RColorBrewer",
              "DT", "kableExtra")
pacman::p_load(char = packages)
```

```{r}
## Script with functions
source(here("scripts/report_funs.R"))
tmpl_file <- here("reports/diffabund_template.Rmd")
```

```{r}
## Define input files
ps_file <- here("results/phyloseq/ps_dadatax_filt.rds")
```

```{r}
## DATASET-SPECIFIC SETTINGS
trt_lev_deseq <- c("LS", "L", "S")   # Deseq would replace the `+` by a `.`
trt_lev_mod <- c("L+S", "L", "S")
trt_lev_plot <- c("L", "S", "L+S")
trt_cols <- brewer.pal(length(trt_lev_plot), "Set1")
names(trt_cols) <- trt_lev_plot
focal_var <- "treatment_deseq"       # Focal variable

## Prevalence thresholds - features that do not pass these thresholds are excluded from for statistical testing
## 10% threshold from https://www.biorxiv.org/content/10.1101/2021.05.10.443486v1.full
PREV_FRAC <- 0.1         # Fraction of samples the ASV/pathway should be present in at prevalence `PREV_N`
PREV_N <- 1              # Min count for the ASV/pathway for at least `PREV_FRAC` of the samples
prev_fun <- function(x) sum(x > PREV_N) > (PREV_FRAC * length(x))
```

```{r}
## General settings
set.seed(100)
theme_set(theme_bw(base_size = 13))
```

```{r}
## Load phyloseq object
ps <- readRDS(ps_file)
```

```{r}
## Remove sample with fewest count, which is the only in the NC treatment
smp_least <- names(sample_sums(ps)[sample_sums(ps) == min(sample_sums(ps))])
ps_min1 <- subset_samples(ps, sampleID != smp_least)
```

```{r}
## Set focal phyloseq object
ps <- ps_min1
```

```{r}
## Metadata
meta <- meta(ps) %>%
  mutate(treatment_deseq = factor(sub("\\+", "", treatment), trt_lev_deseq),
         treatment = factor(treatment, levels = trt_lev_mod),
         treatment_plot = factor(treatment, levels = trt_lev_plot))
```

```{r}
## Prepare a list with data for a focal taxonomic level
prep_taxrank <- function(taxrank) {
  
  if (taxrank != "ASV") ps <- tax_glom(ps, taxrank = taxrank, NArm = TRUE)

  counts <- t(otu_table(ps)) %>% as(., "matrix") %>% round()
  counts <- counts[apply(counts, 1, prev_fun), ]
  
  dds <- mk_deseq(counts, meta)
  dds <- estimateSizeFactors(dds)
  
  counts_norm <- sweep(counts, 2, sizeFactors(dds), "/")
  
  tax <- as.data.frame(tax_table(ps)) %>%
    rownames_to_column("ID") %>%
    dplyr::select(-domain) %>% 
    filter(ID %in% row.names(counts))

  meta <- meta
  
  return(list(ps = ps, dds = dds, counts_norm = counts_norm,
              tax = tax, meta = meta))
}
```

```{r}
## Running DESeq at the ASV level returns an error ('every gene contains at least one zero')
## Therefore, we first filter the data to make sure this is no longer the case
prep_taxrank_asv <- function() {
  counts <- t(otu_table(ps)) %>% as(., "matrix") %>% round()
  counts <- counts[apply(counts, 1, prev_fun), ]
  
  asv_w_least_miss <- names(sort(apply(counts, 1, function(x) sum(x == 0))))[1]
  samp_miss <- which(counts[asv_w_least_miss, ] == 0)
  counts2 <- counts[, -samp_miss]
  meta2 <- meta %>% filter(! sampleID %in% names(samp_miss))
  dds <- mk_deseq(counts2, meta2)
  dds <- estimateSizeFactors(dds)
    
  counts_norm <- sweep(counts2, 2, sizeFactors(dds), "/")
    
  tax <- as.data.frame(tax_table(ps)) %>%
    rownames_to_column("ID") %>%
    dplyr::select(-domain) %>%
    filter(ID %in% row.names(counts))
  
  taxdat <- list(ps = ps, dds = dds, counts_norm = counts_norm,
                 tax = tax, meta = meta2)
  return(taxdat)
}
```

```{r}
## Process DA results table
process_res <- function(res, taxrank) {
  if (taxrank != "ASV") res <- res %>% dplyr::select(-ID)
  if (taxrank == "genus") res <- res %>% dplyr::select(-species)
  if (taxrank == "family") res <- res %>% dplyr::select(-genus, -species)
  if (taxrank == "order") res <- res %>% dplyr::select(-genus, -species, -family)
  if (taxrank == "class") res <- res %>% dplyr::select(-genus, -species, -family, -order)
  return(res)
}
```

```{r}
## Create a report section from template file - ASV level
make_section <- function(taxrank, lev1, lev2,
                         var = focal_var,
                         x_var = "treatment", col_var = "treatment") {
  
  src <- knitr::knit_expand(file = tmpl_file,
                            taxrank = taxrank,
                            lev1 = lev1, lev2 = lev2,
                            var = var, x_var = x_var, col_var = col_var)
  
  res <- knitr::knit_child(text = src, quiet = TRUE)
  
  cat(res, sep = "\n")
}
```

<br>

------

## Methods

### Summary for manuscript

Differential abundance analysis was performed using the R/Bioconductor package
`DESeq2` [@love_moderated_2014] at the ASV, genus, family, order, and class level.

Differential abundance analysis was only performed for individual features
present in at least 10% of the samples in a given comparison,
following the recommendation in @nearing_microbiome_2021.
Reported p-values were adjusted using the Benjamini-Hochberg multiple-testing
correction method,
and we used a significance threshold of 0.05;
for models with pairwise comparisons,
we additionally required an effect size of 1 to consider an effect significant.
ASV counts of differentially abundant taxa were visualized after normalization
with the `ALDEx2` function `aldex.clr()` for consistency with the
differential abundance analysis.

### Additional notes

- I am reporting all features with a (Benjamini-Hochberg) multiple-testing
  corrected p-value below 0.1.
  
- A heatmap for a given comparison is only shown if at least two features
  have a corrected p-value below 0.1.

- **DESeq can be sensitive to outliers (when there are 2 or more) and**
  **looking at the boxplots, **
  **I think most of the significantly differentially abundant ASVs don't look**
  **too exciting, but there seem to be are some more genuine differences at**
  **higher taxonomic levels.**
  
- Related to the point above: I also use an alternative pacakge for differential
  abundance analysis, `ALDEx2`, which is less sensitive to outliers and
  generally more conservative. If need be, I could run that too, for this data.

<br>

------

## ASV level

```{r, results='asis'}
taxrank <- "ASV"
make_section(taxrank, "L", "LS")
make_section(taxrank, "S", "LS")
```

## Genus level

```{r, results='asis'}
taxrank <- "genus"
make_section(taxrank, "L", "LS")
make_section(taxrank, "S", "LS")
```

## Family level

```{r, results='asis'}
taxrank <- "family"
make_section(taxrank, "L", "LS")
make_section(taxrank, "S", "LS")
```

## Order level

```{r, results='asis'}
taxrank <- "order"
make_section(taxrank, "L", "LS")
make_section(taxrank, "S", "LS")
```

## Class level

```{r, results='asis'}
taxrank <- "class"
make_section(taxrank, "L", "LS")
make_section(taxrank, "S", "LS")
```

## References
