---
title: |
  | Exploratory analysis -- <br> Salmonella project metabarcoding data
pagetitle: "Exploratory analysis"
author: "Jelmer Poelstra (poelstra.1@osu.edu), MCIC Wooster"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: cerulean
    highlight: kate
    toc: true
    toc_float: true
    fig_caption: true
    anchor_sections: true
    df_print: kable
    css: html_page.css
bibliography: salmonella.bib
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  eval = TRUE,
  cache = FALSE,
  warning = FALSE,
  message = FALSE,
  out.width = "90%"
)
knitr::opts_knit$set(root.dir = here::here(),
                     output.dir = here::here("reports/html/"))
```

```{r}
## Load packages
if (!"pacman" %in% installed.packages()) install.packages("pacman")
packages <- c("tidyverse", "here",
              "plotly", "pheatmap", "patchwork", "randomcoloR",
              "DT", "kableExtra",
              "vegan", "phyloseq",
              "DESeq2", "microbiome", "metagenomeSeq",
              "remotes")
pacman::p_load(char = packages)
```

```{r}
## Source script with functions
source(here("scripts/report_funs.R"))
```

```{r}
## Define input files
indir <- here("results/phyloseq/")
ps_file <- file.path(indir, "ps_dadatax_filt.rds")
```

```{r}
## DATASET-SPECIFIC SETTINGS
trt_levels <- c("NC", "L", "S", "L+S")
trt_cols <- brewer.pal(length(trt_levels), "Set1")
names(trt_cols) <- trt_levels
```

```{r}
## General settings
set.seed(100)
theme_set(theme_bw(base_size = 13))
```

```{r}
## Load phyloseq objects
ps <- readRDS(ps_file)

## Treatment factor levels
sample_data(ps)$treatment <- factor(sample_data(ps)$treatment, levels = trt_levels)
```

```{r}
## Phyloseq object for ordination visualization
ps_prop <- transform(ps, "compositional")
```

```{r}
## Extract parts of the phyloseq object
meta <- meta(ps)

tax <- as.data.frame(tax_table(ps)) %>%
  rownames_to_column("ASV") %>%
  mutate(across(everything(), ~ str_trunc(.x, width = 25))) # Avoid very long taxon names

counts_prop <- otu_table(ps_prop) %>% as.data.frame() %>% t()
counts <- otu_table(ps) %>% as.data.frame() %>% t()

dds <- mk_deseq(counts, meta)
dds <- estimateSizeFactors(dds)
counts_norm <- sweep(counts, 2, sizeFactors(dds), "/")
```

```{r}
## Mean frequencies and counts for each ASV
freqs_prop <- data.frame(
  ASV = taxa_names(ps_prop),
  mean_prop = round(taxa_sums(ps_prop) / nsamples(ps_prop), 5),
  mean_count = round(taxa_sums(ps) / nsamples(ps))
  )
```

<br>

----

## Methods

### Summary for manuscript

The R/Bioconductor package `phyloseq` (version 1.38.0, @mcmurdie_phyloseq_2013)
was used to store the resulting count table, taxonomy table, phylogenetic tree,
and a metadata table as a single R object and perform a number of downstream analysis,
such as calculating weighted UniFrac distances among samples,
performing a Principal Coordinate Analysis (based on UniFrac distances),
and agglomerating counts at higher taxonomic levels.

We examined overall differences in microbiome composition between treatments
using a PERMANOVA with the weighted UniFrac distance between samples as the
responding variable.
This was done with the `adonis2()` function from the R package `vegan`
(version 2.5.7, @oksanen_vegan_2020).

<br>

------

## Ordination (PCoA)

```{r}
ord <- ordinate(ps_prop, method = "PCoA",
               distance = phyloseq::distance(ps_prop, method = "wunifrac"))
```

### Interactive plots {.tabset .tabset-fade .tabset-pills}

These plots are interactive so you can (1) hover over points and see more info,
and (2) remove certain categories of point by clicking on the legend.

#### Axis 1 & 2

```{r}
p <- plot_ordination(ps_prop, ord, color = "treatment") +
  geom_point(aes(text = paste("Sample ID:", sampleID, "\n",
                              "Treatment:", treatment, "\n")),
             size = 2) +
  scale_color_manual(values = trt_cols)
ggplotly(p, tooltip = "text")
```

#### Axis 3 & 4

```{r}
p <- plot_ordination(ps_prop, ord, color = "treatment", axes = c(3, 4)) +
  geom_point(aes(text = paste("Sample ID:", sampleID, "\n",
                              "Treatment:", treatment, "\n")),
             size = 2) +
  scale_color_manual(values = trt_cols)
ggplotly(p, tooltip = "text")
```

### Static plots {.tabset .tabset-fade .tabset-pills}

#### Axis 1 & 2

```{r}
plot_ordination(ps_prop, ord, color = "treatment") +
  geom_point(size = 2.5) +
  scale_color_manual(values = trt_cols) +
  theme(panel.grid.minor = element_blank())
```

#### Axis 3 & 4

```{r}
plot_ordination(ps_prop, ord, color = "treatment", axes = c(3, 4)) +
  geom_point(size = 2.5) +
  scale_color_manual(values = trt_cols) +
  theme(panel.grid.minor = element_blank())
```

### Plotting ASVs instead of taxa

```{r}
colorby_columns <- which(colnames(tax_table(ps)) == "phylum")
ncols <- length(unique(tax_table(ps)[, colorby_columns]))
mycols <- distinctColorPalette(ncols)

pdf <- plot_ordination(ps_prop, ord, type = "taxa", justDF = TRUE) %>%
  rownames_to_column("ASV") %>%
  merge(., freqs_prop, by = "ASV")

p <- ggplot(pdf) +
  aes(x = Axis.1, y = Axis.2, color = phylum, size = mean_prop) +
  geom_point(aes(text = paste("ASV:", ASV, "\n",
                              "Phylum:", phylum, "\n",
                              "Class:", class, "\n"))) +
  scale_color_manual(values = mycols) +
  labs(size = "Abundance")
ggplotly(p, tooltip = "text")
```

<br>

-----

## Permanova  {.tabset .tabset-fade .tabset-pills}

```{r}
get_perma <- function(ps, treatments) {
  
  fps <- prune_samples(sample_data(ps)[["treatment"]] %in% treatments, ps)
  
  perma_res <- adonis2(
    UniFrac(fps, weighted = TRUE) ~ treatment,
    data = as(sample_data(fps), "data.frame"),
    permutations = 100000
    )

  perma_res %>% kable() %>% kable_styling(full_width = FALSE)
}
```

### Across all treatments

```{r}
get_perma(ps_prop, treatments = trt_levels)
```

<br>

-----

### `NC` vs `S`

```{r}
get_perma(ps_prop, treatments = c("NC", "S"))
```

<br>

-----

### `NC` vs `L`

```{r}
get_perma(ps_prop, treatments = c("NC", "L"))
```

<br>

-----

### `S` vs `L+S`

```{r}
get_perma(ps_prop, treatments = c("S", "L+S"))
```

<br>

-----

<br>

-----

## Abundance by taxonomic level {.tabset .tabset-fade .tabset-pills}

```{r}
bar_group <- function(fps, taxrank, cols) {
  means <- psmelt(fps) %>%
    group_by(.data[[taxrank]], treatment) %>%
    summarize(mean_abund = mean(Abundance))
  
  ggplot(means) +
    aes(x = treatment, y = mean_abund, fill = .data[[taxrank]]) +
    geom_col() +
    scale_y_continuous(expand = expansion(mult = c(0, 0.01)),
                       labels = scales::label_percent()) +
    scale_fill_manual(values = cols) +
    labs(y = "Mean abundance") +
    theme(panel.grid.major.x = element_blank(),
          panel.grid.minor = element_blank())
}
```

```{r}
bar_indiv <- function(fps, taxrank, cols) {
  plot_bar(fps, fill = taxrank) +
    facet_grid(cols = vars(treatment), scales = "free_x", space = "free") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.01)),
                       labels = scales::label_percent()) +
    scale_fill_manual(values = cols) +
    labs(x = NULL) +
    theme(axis.text.x = element_text(angle = 270),
          panel.grid.major.x = element_blank(),
          panel.grid.minor = element_blank())
}
```

### Phylum {.tabset .tabset-fade .tabset-pills}

#### By sample

```{r}
ps_foc <- tax_glom(ps_prop, taxrank = "phylum", NArm = FALSE)
cols <- distinctColorPalette(ntaxa(ps_foc))
```

```{r}
bar_indiv(ps_foc, "phylum", cols)
```

#### Grouped

```{r}
bar_group(ps_foc, "phylum", cols)
```

#### Without _Firmicutes_ -- by sample

```{r}
ps_foc <- tax_glom(ps_prop, taxrank = "phylum", NArm = FALSE)
ps_foc <- subset_taxa(ps_foc, phylum != "Firmicutes" | is.na(phylum))
cols <- distinctColorPalette(ntaxa(ps_foc))
```

```{r}
bar_indiv(ps_foc, "phylum", cols)
```

#### Without _Firmicutes_ -- grouped

```{r}
bar_group(ps_foc, taxrank = "phylum", cols)
```

### Class {.tabset .tabset-fade .tabset-pills}

#### By sample

Only showing classes with a mean proportional abundance higher than 0.01.

```{r}
ps_foc <- tax_glom(ps_prop, taxrank = "class", NArm = FALSE)
ps_foc <- subset_taxa(ps_foc, (taxa_sums(ps_foc) / nsamples(ps_foc)) > 0.01)
cols <- distinctColorPalette(ntaxa(ps_foc))
```

```{r}
bar_indiv(ps_foc, "class", cols)
```

#### Grouped

```{r}
bar_group(ps_foc, "class", cols)
```

#### Without _Clostridia_ -- by sample

Showing all classes except _Clostridia_.

```{r}
ps_foc <- tax_glom(ps_prop, taxrank = "class", NArm = FALSE)
ps_foc <- subset_taxa(ps_foc, class != "Clostridia" | is.na(class))
cols <- distinctColorPalette(ntaxa(ps_foc))
```

```{r}
bar_indiv(ps_foc, "class", cols)
```

#### Without _Clostridia_ -- grouped

```{r}
bar_group(ps_foc, "class", cols)
```

### Order {.tabset .tabset-fade .tabset-pills}

Only showing orders with a mean proportional abundance higher than 0.01.

```{r}
ps_foc <- tax_glom(ps_prop, taxrank = "order", NArm = FALSE)
ps_foc <- subset_taxa(ps_foc, (taxa_sums(ps_foc) / nsamples(ps_foc)) > 0.01)
cols <- distinctColorPalette(ntaxa(ps_foc))
```

#### By sample

```{r}
bar_indiv(ps_foc, "order", cols)
```

#### Grouped

```{r}
bar_group(ps_foc, "order", cols)
```

### Family {.tabset .tabset-fade .tabset-pills}

Only showing families with a mean proportional abundance higher than 0.01.

```{r}
ps_foc <- tax_glom(ps_prop, taxrank = "family", NArm = FALSE)
ps_foc <- subset_taxa(ps_foc, (taxa_sums(ps_foc) / nsamples(ps_foc)) > 0.01)
cols <- distinctColorPalette(ntaxa(ps_foc))
```

#### By sample

```{r}
bar_indiv(ps_foc, "family", cols)
```

#### Grouped

```{r}
bar_group(ps_foc, "family", cols)
```

### Genus {.tabset .tabset-fade .tabset-pills}

Only showing genera with a mean proportional abundance higher than 0.1.

```{r}
ps_foc <- tax_glom(ps_prop, taxrank = "genus", NArm = FALSE)
ps_foc <- subset_taxa(ps_foc, (taxa_sums(ps_foc) / nsamples(ps_foc)) > 0.01)
cols <- distinctColorPalette(ntaxa(ps_foc))
```

#### By sample

```{r}
bar_indiv(ps_foc, "genus", cols)
```

#### Grouped

```{r}
bar_group(ps_foc, "genus", cols)
```

<br>

------

## Focal taxa

### Salmonella

Not present in the dataset -- in fact, the phylum _Pseudomonadota_ is not found
in the dataset at all.

```{r}
#tax <- as.data.frame(ps@tax_table)
#any(tax$phylum == "Pseudomonadota")
#unique(tax$phylum)
```

### Lactobacillus  {.tabset .tabset-fade .tabset-pills}

The following taxa in the Order _Lactobacillales_ were detected:

#### Table

```{r}
lacto_tax <- tax %>% filter(order == "Lactobacillales")
```

```{r}
make_dt(lacto_tax, simple_mode = TRUE)
```

#### Heatmap

```{r}
mycols <- list(trt_cols)
names(mycols) <- "treatment"
  
plot_heatmap(lacto_tax$ASV,
             count_mat = counts_norm, meta_df = meta,
             groups = c("treatment"), annot_colors = mycols)
```

#### Boxplots

```{r, out.width = "80%"}
walk(lacto_tax$ASV, plot_abund,
    count_mat = counts_norm, tax_df = tax, meta = meta,
    tax_df_column = "ASV")
```

<br>

------

## Most abundant ASVs {.tabset .tabset-fade .tabset-pills}

In the plot and tables below,
the "mean proportion" / `mean_prop` of an ASV refers to the mean proportion
of all ASVs detected in each sample that are assigned to that ASV
(e.g., if a sample has 1000 ASVs and 100 of them are ASV_1, then ASV_1 has
a proportion of 0.1 in that sample).

### Abundance heatmap with top-20

Showing the top-20 most abundant ASVs,
sorted by abundance & starting with the most abundant.
In the `Table with top-50` tab, you can see which taxa these are.

```{r}
topASV_df <- merge(freqs_prop, tax_table(ps_prop),
                   by.x = "ASV", by.y = "row.names") %>%
  select(-domain) %>% 
  arrange(desc(mean_prop))
```

```{r, out.width="80%", fig.height=7}
mycols <- list(trt_cols)
names(mycols) <- "treatment"
  
plot_heatmap(IDs = topASV_df$ASV[1:20],
             meta_df = meta, count_mat = counts_norm,
             groups = "treatment", annotation_colors = mycols,
             log_transform = TRUE)
```

### Table with top-50

Containing the top-50 most abundant ASVs.

<br>

```{r}
make_dt(topASV_df[1:50, ], pageLength = 15)
```

### ASV abundance distribution

In the plot below, the inset is zoomed-in on the left hand portion of the plot.

```{r, out.width="90%"}
p1 <- ggplot(data = freqs_prop, aes(x = mean_prop)) +
  geom_histogram(bins = 500) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = expansion(mult = c(0, .05))) +
  labs(x = "Mean proportion of an ASV", y = "Number of distinct ASVs")

p2 <- ggplot(data = freqs_prop, aes(x = mean_prop)) +
  geom_histogram(bins = 500) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) +
  coord_cartesian(xlim = c(0, 0.015)) + #, ylim = c(0, 50)) +
  labs(x = "Mean proportion of an ASV", y = "Number of distinct ASVs") +
  theme_bw(base_size = 11) +
  theme(plot.background = element_rect(fill = "grey90"),
        panel.background = element_rect(fill = "grey90"),
        plot.margin = margin(10, 20, 10, 10, unit = "pt"))

p1 + patchwork::inset_element(p2, 0.4, 0.4, 0.98, 0.99)
```

<br>

----

## ASVs by taxonomic group {.tabset .tabset-fade .tabset-pills}

```{r, out.width="85%"}
plot_asv_cnt <- function(tax_lev, tax_df) {
  
  ## Determine plotting order
  ord <- tax_df %>%
    dplyr::count(.data[[tax_lev]], sort = TRUE) %>%
    pull(.data[[tax_lev]])
  
  tax_df %>% 
    mutate(tax_foc = factor(.data[[tax_lev]], levels = ord)) %>% 
    ggplot(aes(x = tax_foc)) +
    geom_bar() +
    scale_y_continuous(expand = expansion(mult = c(0, .05))) +
    labs(x = tax_lev) +
    theme(axis.text.x = element_text(angle = 270, hjust = 0, vjust = 0.5))
}
```

Barplots showing the number of distinct ASVs in each taxonomic group.

### Phylum

```{r, out.width="85%"}
plot_asv_cnt(tax_lev = "phylum", tax_df = tax)
```

### Class

```{r, out.width="85%"}
plot_asv_cnt(tax_lev = "class", tax_df = tax)
```

### Order

```{r, out.width="85%"}
plot_asv_cnt(tax_lev = "order", tax_df = tax)
```

### Family

```{r, out.width="85%"}
plot_asv_cnt(tax_lev = "family", tax_df = tax)
```

<br>

----

## References
